1.0.0 Refactor (Issue #36)

Lindsay-Needs-Sleep opened the pull request:
I am excited to present to you the 1.0.0 re-write of the plugin!  The plugin now follows the chrome.cast api much more closely!  And (as far as I am aware) works without issue!

I say re-write because I estimated that ~80% of Android was changed.  And iOS was moved to objc to fit the cordova plugin standard (so a ~100% rewrite).

There should be very few changes required to migrate from 0.x to 1.0.0.  Mostly if you depended on `getRouteListElement` you will have to switch to `scanForRoutes`/`stopScan`, or if you depended on behavior not found on chrome desktop's api implementation.

These changes have been tested on:
- Android 4.4, 7.x
- iOS 9.3 (iPad 3), 10.3 (iPhone 5), 12.4.2 (iPad mini 2), 13.2.2 (emulator iPad 7, emulator iPhone 8)
- Chrome desktop browser 78.0 (the tests pass on this version of chrome, which demonstrates that the same code will work on all 3 platforms [basically])

The largest changes of this refactor are:

- [x] iOS, Android, and Chrome Desktop all work the same now!  (With a few caveats.)
All 3 will pass the same set of tests included in the plugin.

- [x] iOS and Android have `scanForRoutes`, `stopScan`, and `selectRoute` to replace `getRouteListElement`

- [x] iOS and Android still do not implement the full chromecast API, but a few more have been added, and existing ones have been made to match chrome desktop behavior.

**Bug Fixes**: 
- Fixes #17 (cancel should return error)
- Fixes #22 for Android (getRouteListElement() returns empty object) non-issue now
- Fixes #22 for iOS (getRouteListElement() returns empty object) non-issue now
- Fixes #23 Metadata not set
- Fixes #26 #27 (compile/dependancy error)
- Fixes #32 (duplicate routes are displayed)
- Fixes #36 (Because that is this issue :p)
- Fixes #43 (crash on requestSession)
- Fixes #41 remove deprecated `sendJavascript` usage
- Fixes #47 iOS crash
- Fixes #48 selectRoute crash
- Fixes #49 selectRoute silent fail
- Fixes #50 Switch to Mocha for tests
- Play audio files https://github.com/miloproductionsinc/cordova-plugin-chromecast/issues/3

**Improvements:**
- Upgraded to CAF v3, this got rid of a huge amount of deprecation notices.
- Added eslint tests to ensure uniform js code stlying
- Added a style checker for java
- Updated documentation on: how to run style checkers, how to run tests.
- Added a pull request template which instructions prospective pull requesters to run eslint and the tests
- Repaired and added to the existing tests enforcing chrome desktop SDK behavior.
- On app initialize you will receive any previously started+active session.  (This includes navigating to a new page, or force killing the app and re-loading.)
- Implemented startRouteScan and stopRouteScan to replace getRoutes ([see why](https://github.com/jellyfin/cordova-plugin-chromecast/issues/22#issuecomment-530773677))
- fixed selectRoute for Android 4.4.2
- Allow native error responses to send a json object "code" and "description" parameters in addition to just the error code string. 
- Implement basic queue functionality
- Tests can run on chrome desktop browser as well (with a couple stubs to enable missing functionality)
  - This is super useful because you can write tests and check that they work correctly on chrome first.  Which is very important when chrome's implementation is what we are implementing.
- Added a basic code example
- Updated readme with all kinds of instructions (differences between chrome, api list, plugin-specific api list, api usage example, support+quirks, setup, how to test, how to run formatting test)
- Added pull request template
- Added scripts to assist in enforcing code formatting for JS and Java

**Auto-tests**:
- initialize
    - receiverListener
    - sessionListener
- leaveSession
- stopSession
- loadMedia Video
- media.setVolume
- media.play
- media.pause
- media.stop
- session.setReceiverVolumeLevel
- session.setReceiverMuted
- requestSession
- cordova.startRouteScan
- cordova.stopRouteScan
- cordova.selectRoute
- MediaMetadata (All types with most of their special params)
- loadMedia Audio (Did not test playback controls)
- loadMedia Image
- load queue
- jump to queue item

**Manual Tests**:
- Start session, reload page, call initialize, do we get session again?
- Start session, force kill app, restart app, do we get session again?
- Start session, join session with additional device, **leave** session on original, session still continues?, force kill and restart original device, original device should not rejoin the session automatically.
- for session Leave: probably need to join the session from an additional device,then leave the session, restart app, does it rejoin the session? (it shouldn't) (pre-req rejoin on restart must work)
- for loadmedia from external sender: start session on device, join session on other device, loadmedia from other device, is session.addMediaListener triggered?
- for external exit: Start session, kill session with other device, requestSession should show the the join dialog, not stopCasting

In addition to all the issues listed above, this probably fixes these as well:
- #52 
- #53

Sel-en-ium committed:
Move js files to www to follow the standard-ish plugin format

Sel-en-ium committed:
Add eslint test (so that the code can have a unified [standard-ish] style)  (Will satisfy the eslint test in another commit.)

Sel-en-ium committed:
Updated the jasmine tests to use the newer cordova-plugin-test-framework version. (Most tests fail.  Will try to fix in another commit.)

# Conflicts:
#	README.md
#	tests/tests.js

Sel-en-ium committed:
Satisfy ESLint.  (To [commit history], I'm sorry.)

# Conflicts:
#	tests/tests.js
#	www/chrome.cast.js

Sel-en-ium committed:
Update cast library so that it doesn't conflict with the androidx libraries.
Fixes issue#26 https://github.com/jellyfin/cordova-plugin-chromecast/issues/26

Sel-en-ium committed:
Update cast library so that it doesn't conflict with the androidx libraries.
Fixes issue#26 https://github.com/jellyfin/cordova-plugin-chromecast/issues/26

Sel-en-ium committed:
Merge branch 'androidx'

Sel-en-ium committed:
Merge branch 'gitignore' into eslint

Sel-en-ium committed:
Merge branch 'eslint'

Lindsay-Needs-Sleep committed:
Switch to MediaRouteChooserDialog to simplify the dialog building process. (Chromecast.java)
We shouldn't prevent requestSession from going through if there is no receiver.  It handles this fine.  It just shows that it is searching for a chromecast.  This follows the behavior of chrome browser cast SDK. (chrome.cast.js)
Remove unused imports. (Chromecast.java)
Try to be a bit more safe with the threads. (Chromecast.java)
Just pass the Chromecast instance to the callback on initialization.  (Chromecast.java and ChromecastMediaRouterCallback.java)

Lindsay-Needs-Sleep committed:
We should not be logging to javascript all the time.  It is an abuse of power and annoying.  Use the cordova log utility instead.

Lindsay-Needs-Sleep committed:
WIP: Got the test framework to run.  Basically all the tests are broken though.
Updated the README.md with instructions on how to run these tests.
We shouldn't load the tests.js file when the user is using the plugin for production use.
We shouldn't include the test framework as a dependency.  (Even a dependency of the test plugin.)
Added a pull_request_template.md

Lindsay-Needs-Sleep committed:
WIP: Got the first couple tests working.
Managed to match the desktop chrome SDK behavior regarding receiver updates on start up.
Added some additional tests to cover more of the basic situations with receiver updates.
Simplified all the get route stuff (will fix it up properly later)

Lindsay-Needs-Sleep committed:
WIP: Got request session somewhat working
capture routeUnselected reason

Lindsay-Needs-Sleep committed:
WIP: Add a video that will actually play and fix the load media test

Lindsay-Needs-Sleep committed:
.gitignore update

Lindsay-Needs-Sleep committed:
Got all the tests to pass

Lindsay-Needs-Sleep committed:
Merge remote-tracking branch 'remotes/jellyfin/master'

Lindsay-Needs-Sleep committed:
remove receiverAvailable because it does nothing
stop spamming emaiAllRoutes

Lindsay-Needs-Sleep committed:
Version bump to 1.0.0

Lindsay-Needs-Sleep committed:
Should make sure assets are in the plugin folder so that we don't pollute the top level of www

Lindsay-Needs-Sleep committed:
reject and catch don't work great because jasmine does not give very good stack trace.  So remove them and check error as soon as possible to the criminal code.

Lindsay-Needs-Sleep committed:
Update package.json to match licensing

Lindsay-Needs-Sleep committed:
Fix on requestSession cancel handling to match the documentation and chrome desktop chromecast behavior.  We are supposed to receive a callback to the error handler.  Not have it silenced.  This is important if you have started connection animations.  If you never receive a callback you don't know when to stop the animations, and whether to transition back to the unconnected state or to a connected stated.  If you wish to ignore the cancel error, just filter it out in your error callback. eg. if (err.code === 'cancel') { //ignore }

Lindsay-Needs-Sleep committed:
Forgot to remove unused references to receiverAvailable related vars.

Lindsay-Needs-Sleep committed:
Update readme

Lindsay-Needs-Sleep committed:
Integrate receiverListener test into initialize.  Having cross-test dependencies is not great.  Though, we are on our way to a monolith test this way.  :/  Pick our poison I guess.

Lindsay-Needs-Sleep committed:
rename for next commit

Lindsay-Needs-Sleep committed:
Update to cast v3 sender to remove deprecated functions.
Session is automatically "rejoined" when the webview goes to a new page and calls initialize (this follows chromecast desktop sdk)
Do not call sessionListener on requestSession success (this does not follow chromecast sdk desktop behavior)

Lindsay-Needs-Sleep committed:
Added a java file style checker that enforces 4 spaces.
(So I changed all the tabs to spaces in the java files.  Sorry commit history. o.o)
It was a lot of work, so I disable a couple of the java checkstyle rules.  They are listed in the first comment in check_style.xml.
Issue #36 progress

Lindsay-Needs-Sleep committed:
Added stop casting dialog.
finally got the tests to early terminate on error (It's ugly.  We should consider switching to mocha.)
Work done for Issue #36

Lindsay-Needs-Sleep committed:
Apparently my changes for scanning for routes, stopping the scan, and selecting the route do work.  So I am going to commit so that they are in the history.
Also removed the private _sessions array in chrome.cast.js.  The app side can only ever maintain 1 session, so it is strange for the client side to track multiple sessions. (If it does have multiple sessions, all but the latest will be dead to it anyways.
Relevant to Issue #36

Lindsay-Needs-Sleep committed:
fix createAppImagesObject

Lindsay-Needs-Sleep committed:
Merge Lewazo's changes.  Filter out active duplicate routes and duplicate cast group routes.

Lindsay-Needs-Sleep committed:
Added tests for session Leave and session Stop.  Cleaned up their functionality to match chrome desktop SDK a bit more.
Add names to constructors in chrome.cast.js to improve out in test.js test().toBeDefined()

Lindsay-Needs-Sleep committed:
Added package.json script to automatically fix the js errors it can.

Lindsay-Needs-Sleep committed:
Got the plugin to automatically resume sessions after being the app has been force quit and restarted.  This was rough.
Added defaults to _sessionListener and _receiverListener.
Issue #36

Lindsay-Needs-Sleep committed:
Should technically mark the package version as a "dev" version until completely stable.

Lindsay-Needs-Sleep committed:
(android) Make the constants match the desktop chrome SDK.
Issue #36

Lindsay-Needs-Sleep committed:
Fixed all kinds of problems with setVolume (media and receiver)
- following the chrome SDK behavior for setVolume more closely
- need to trigger sessionUpdate on receiver volume change
- added method which accepts Integer for setMediaVolume (in case client puts exactly 0 or 1)
Allowed null arguments to be received on the Android side
- since all our android method entry points accept nullable objects this should be allowed
- in particular, this allows us to have one setMediaVolume (instead of setMediaVolume and setMediaMuted) to accept null args which mean "don't update the level/mute state"
Session handling fixes
- fixed how session.status is added to session
- fixed leave session (was passing true to endSession, oops)
- changed "SESSION_ERROR" instances in android to "session_error" to match the SDK directly
Cleaned up and improved tests
- more tests checked for accurate status updates
- made some tests more readable with callOrder
Cleaning
- removed ChromecastConnection.Callback since it was exactly the same as Runnable, just use that.
Streamlined the session end event callback process
- Chromecast.java has only one location which triggers the client's session update listener (with end status)
- ChromecastSession.java detects external reasons for disconnect with onApplicationDisconnected (eg. timeout and someone else stealing the cast)
- ChromecastConnection.java detects self-triggered reasons for disconnect (eg. sessionEnd, sessionLeave)
  (in which case it notifies ChromecastSession.java that a session has ended, which then tells Chromecast.java)
Simplified receiverListener
- just pass a boolean to _receiverUpdate
Untested fix:
- maybe fixed the message functions and updates (needs a test)

Issue #36 (1.0.0 development)

Lindsay-Needs-Sleep committed:
Got a remote video for testing instead!  Wonderful!  Thanks @anthonylavado!
Issue #36

Lindsay-Needs-Sleep committed:
SessionManager.startSession did not work reliably on android 4.4.2, so switch to MediaRouter.selectRoute()
Add isNearbyDevice to routes output from startRouteScan (mostly so testing can avoid attempting to join these routes, since they require manual input.  But could be useful to the user as well.)
Remove unnecessary chrome.cast.js global vars.
Added polyfill for Promise to tests since Android 4.4.3- does not have Promise natively.
Contributing to Issue #36
Relevant to Issue #22

Lindsay-Needs-Sleep committed:
Merge remote-tracking branch 'remotes/jellyfin/master' into 1.0.0
Merge pull request #39 from lewazo/fix/json-escaping
Fix the exception encountered when playing media with single quote character in the metadata

Lindsay-Needs-Sleep committed:
Fixed up mediaLoaded.  It should be called whenever a new media is loaded, even by an external sender.
Towards issue #36

Lindsay-Needs-Sleep committed:
_mediaUpdated event is already fired by Media.prototype._update, so we are emitting twice each update.  In the case that we are creating a new media object, it is impossible for there to be any listeners on it yet, so no point in emitting.
Towards issue #36

Lindsay-Needs-Sleep committed:
Only ChromecastSession needs to hold a reference to session.
Since we are keeping scanForRoutes, might as well use it while checking for available receivers.
Make the JSON creator functions static so they aren't state dependent.
Issue #36

Lindsay-Needs-Sleep committed:
(android) (Refactor) Move JSONObject creator functions to utilities.

Lindsay-Needs-Sleep committed:
(android) (Refactor) Decouple the 3 way dependency between Chromecast.java, ChromecastConnection.java, and ChromecastSession.java a little bit.
Move more JSON creation methods to utilities (createSessionObject with state, and createRoutesArray).
connection.selectRoute does not use routeName anymore.
Chromecast.java shouldn't know about CastSession if we can avoid it.

Lindsay-Needs-Sleep committed:
Merge remote-tracking branch 'remotes/jellyfin/master' into 1.0.0

Lindsay-Needs-Sleep committed:
Improve selectRoute with some error handling.
Add isCastGroup to routes returned to startRouteScan.
Relevant to Issue #22
Improvement for issue #36

Lindsay-Needs-Sleep committed:
Improve loadMediaVideo test
Issue #36

Lindsay-Needs-Sleep committed:
Add test to check for video finished state

Lindsay-Needs-Sleep committed:
(android) Remove sendJavascript and replace with an event callback that is set during setup.
Part of issue #36
Fixes issue #41 for android

Lindsay-Needs-Sleep committed:
Update package script to have the standard "npm test"

Lindsay-Needs-Sleep committed:
Move the Api isAvailable check to a central location

Lindsay-Needs-Sleep committed:
(android) Add some notes and error detection for Android 4.4 mysterious, intermittent failure to create a session during selectRoute.

Lindsay-Needs-Sleep committed:
Add check to make sure initialize has been successfully called before most functions are able to work.

Lindsay-Needs-Sleep committed:
Should make sure _session exists before trying to use it since this event can be called at any time

Lindsay-Needs-Sleep committed:
Improvements for selectRoute and requestSession.
Make unique callbacks for selectRoute and requestSession.
Enable retrying to join a route for selectRoute under certain situations.  See issues:
https://github.com/jellyfin/cordova-plugin-chromecast/issues/49
https://github.com/jellyfin/cordova-plugin-chromecast/issues/48

Lindsay-Needs-Sleep committed:
Improve error handling to allow native responses to send a JSONObject with "code" and "description" parameters instead of just the error code string.
Make selectRoute use the full error object.
Towards issue #36

Lindsay-Needs-Sleep committed:
Protect against user passing null as _receiverListener or _sessionListener

Lindsay-Needs-Sleep committed:
Improved startRouteScan/stopRouteScan to handle multiple calls better.
startRouteScan also has it's error callback called with a 'cancel' error when the scan is stopped.
update documentation
Issue #36

Lindsay-Needs-Sleep committed:
(android) Remove e.printStackTrace when building JSONObjects.
Since we expect those exceptions to be thrown often they are basically just annoying.  If a user has experiences unexpected responses they won't help immediately anyways (since there are so many and so frequent).

Lindsay-Needs-Sleep committed:
Kill any route scan that may be going on when setup is called.
Do this because a route scan may have been triggered, then the user navigated away from the page before stopRouteScan was called.
The new page load will call setup and stop any scan.
Issue #36

Lindsay-Needs-Sleep committed:
Only update session.addMediaListnener (MEDIA_LOADED event) when an external sender loads media.
Handle PLAYER_STATE_LOADING, returning buffering is better than unknown.
Remove calls to super because they do nothing.
session.getApplicationMetadata() throws IllegalStateException if called before connected, catch it like the others.

Lindsay-Needs-Sleep committed:
Issue #50 Issue #36 Converted the existing auto tests to mocha and misc improvements/fixes.

Fixed some bugs:
    - typo in ChromecastUtilities "CANCELED"
    - session building, in particulate the media array was not being built correctly

Improvements:
    - Improved many of the tests to be more specific and encourage following desktop chrome behavior more closely
    - Update the readme with more accurate information/documentation/and info about Mocha tests
    - Improved the utils functions (callOrder and waitForAllCalls)
    - Added pre-checks to match chrome behavior.  (Checks for existing session and valid sessionId)
    - followed the _update pattern completely in chrome.cast

Lindsay-Needs-Sleep committed:
Added ability to run tests on chrome to confirm that functionality matches chrome desktop behavior.

Lindsay-Needs-Sleep committed:
Fix chai.js, fix typos in readme, fix css import, rename runner since it is actually just the custom reporter

Lindsay-Needs-Sleep committed:
Fixed up Metadata, added some tests.
Fix Issue #23 for Android and part of https://github.com/miloproductionsinc/cordova-plugin-chromecast/issues/3
Also some better null checking.
And better images handling.
Add RepeatMode to media output.
Issue #36

Lindsay-Needs-Sleep committed:
Issue #23 Fix up the metadata to handle more than strings (ChromecastSession and ChromecastUtilities).
Provide better translations between desktop chrome metadata names and the android names (ChromecastUtilities).
Add tests for loading audio and images.
Should manually trigger media update event after media load because sometimes the MEDIA_UPDATE event fires before the client is able to get a reference to media to addUpdateListener.

Lindsay-Needs-Sleep committed:
Include Mocha fix PR #4051 because I just tried it out and it's nice (fixes a bug with filtering by passes/failures).

Lindsay-Needs-Sleep committed:
Fix tests html, and improve utils output of callOrder functions a bit

Lindsay-Needs-Sleep committed:
Fix crash because of NullPointerException. https://github.com/miloproductionsinc/cordova-plugin-chromecast/issues/3

Lindsay-Needs-Sleep committed:
Added base queue functions.  Towards issue #36
Move some object generation methods to Utilities
Disable method length rule.  We are grown ups and can decide if we want to use ridiculously long methods or not. :)

Lindsay-Needs-Sleep committed:
Improve tests styling a bit

Lindsay-Needs-Sleep committed:
In tests_auto.js improve error output and move setup items to setup.

Lindsay-Needs-Sleep committed:
Added beginning of manual tests (requestSession)
Towards issue #36

Lindsay-Needs-Sleep committed:
Added the restart app / rejoin session manual test
Removed test skipping/memory for manual tests
Move some functions to utils so we can use them from multiple locations

Lindsay-Needs-Sleep committed:
Added more manual Tests.  Added all the major ones where multiple devices interact via the same session.
Towards issue #36

CodeHole7 committed:
finished auto tests

Lindsay-Needs-Sleep committed:
Added a bit more specific instructions to manual tests

CodeHole7 committed:
Auto tests are fixed

CodeHole7 committed:
manual and auto tests completed

CodeHole7 committed:
Auto and manual tests completed

CodeHole7 committed:
auto and manual tests are finished

Lindsay-Needs-Sleep committed:
Fix bug.  Js failure if trying to unset media on non-existent session.

Lindsay-Needs-Sleep committed:
Add tests that specifically test that initialize will return any active session on new page loads.  (Or will not return a session that has been left (session.leave).)  Keep track of test progress through page reloads and app restarts via cookies.

Lindsay-Needs-Sleep committed:
Improve manual instructions a bit and fix forgotten closing quotes in instruction.

Lindsay-Needs-Sleep committed:
Apparently ios deletes cookies during app restart. So switch to localstorage to persist the data through the app restart tests.

Lindsay-Needs-Sleep committed:
Move the sessionRejoin event to after the initialize success callback.  The order of events for initialize when a session is present should be:
1) initialize callback success
2) RECEIVER_LISTENER (false)
3) RECEIVER_LISTENER (true)
4) SESSION_LISTENER (session)

Lindsay-Needs-Sleep committed:
Give slightly longer timeout to accommodate slower devices

Lindsay-Needs-Sleep committed:
Fix session being returned twice on app restart / session rejoin

Lindsay-Needs-Sleep committed:
Fix for iOS 9.  iOS 9 does not support [NSTimer scheduledTimerWithTimeInterval].  Also clean up start/stop scan and selectroute.  Make it so that startScan has a hope of sending updates as available routes change.  selectRoute now also actually tries for 15 seconds.

Lindsay-Needs-Sleep committed:
Style changes.  Xcode is determined to do this.

Lindsay-Needs-Sleep committed:
(ios) ensure that the scan is running while attempting to selectRoute.

Lindsay-Needs-Sleep committed:
Make manual test buttons bigger for ios.

Lindsay-Needs-Sleep committed:
Forgot to apply some button ids

Lindsay-Needs-Sleep committed:
Updated readme with api description and example js

Lindsay-Needs-Sleep committed:
Merge remote-tracking branch 'remotes/jellyfin/master' (Jellyfin updates are irrelevant since the code has changed so much.  But just trying to make it easier to merge this code back to jellyfin/master.

# Conflicts:
#	plugin.xml
#	src/android/Chromecast.java

Lindsay-Needs-Sleep committed:
version update 1.0.0

Lindsay-Needs-Sleep committed:
(ios) Added instructions for meeting ios app distribution requirements

Lindsay-Needs-Sleep commented:
It would be appreciated if someone could run the tests on an android device of version 5.x, 6.x, 8.x, or 9.x.

Lindsay-Needs-Sleep committed:
Update google cast library to 4.4.6 which supports iOS 9.0+

Lindsay-Needs-Sleep committed:
(ios) remove unused emitAllRoutes

Lindsay-Needs-Sleep committed:
[non-change] typo and empty lines change

Lindsay-Needs-Sleep committed:
(ios) we don't use sendJavascript anymore

Lindsay-Needs-Sleep committed:
(ios) Move device list building to utilities. Simplify receiverListener methods.

Lindsay-Needs-Sleep committed:
(ios) media must be an array when returned in session

Lindsay-Needs-Sleep committed:
(ios) decrease cast utilties code duplication

Lindsay-Needs-Sleep committed:
(ios) fix startTime and preloadTime, check for invalidTimeInterval and NaN

Lindsay-Needs-Sleep committed:
(ios) Let ChromecastSession handle all the rejoining logic (also change ChromecastSession initialization to happen only ).  Create rety function.  Store appId in user preferences.  In ChromecastSession, make currentSession a private variable.

Lindsay-Needs-Sleep committed:
(ios) Modify the way scans keep track of whether or not to stop a scan.  make scanCommand a private variable.

Lindsay-Needs-Sleep committed:
(ios) [no change] Remove sendScan (forgot to remove on commit "(ios) Move device list building to utilities.")

Lindsay-Needs-Sleep committed:
(ios) [no change] Remove createSessionObject (the one with no "status" parameter) (forgot to remove on commit "(ios) Let ChromecastSession handle all the rejoining logic...")

Lindsay-Needs-Sleep committed:
(ios) create one endSession function

Lindsay-Needs-Sleep commented:
Don't bother testing right now!  On hold until I fix https://github.com/miloproductionsinc/cordova-plugin-chromecast/issues/5.

Lindsay-Needs-Sleep committed:
(ios) try to join selected route for 5 seconds

Lindsay-Needs-Sleep committed:
(ios) WIP remove isAlive param, it is never used in these contexts, just passed around

Lindsay-Needs-Sleep committed:
(ios) Add CDVPlugin event onReset so that we can stop scans from running when we have changed/refreshed the page

Lindsay-Needs-Sleep committed:
(ios) WIP fix requestSession bug where it was not getting the devices (because of commit "Move device list building to utilities. Simplify receiverListener methods.")

Lindsay-Needs-Sleep committed:
(test) Ensure that chromecast is initialized in before of describe test block

Lindsay-Needs-Sleep committed:
(ios) WIP Calculate the session's status except for the case when we are doing session.leave

Lindsay-Needs-Sleep committed:
(ios) WIP Added endSessionWithCallback to reduce code duplication by allowing requestSession's stop casting option to use the same endSesssion controls

Lindsay-Needs-Sleep committed:
(test) WIP session.stop should happen in a specific order

Lindsay-Needs-Sleep committed:
(test) WIP Remove some unintended copy pasta

Lindsay-Needs-Sleep committed:
(test) (ios) WIP ensure there is no idelReason when media is active

Lindsay-Needs-Sleep committed:
(ios) WIP remove isRequesting and create slighlty unified requestDelegate creators to avoid code duplication

Lindsay-Needs-Sleep committed:
(ios) WIP fix/unify setMediaVolumeWithCommand so that we have just one function for simplicity and so that it does not return until both the volume and mute state have been updated

Lindsay-Needs-Sleep committed:
(ios) WIP Add createLoadMediaRequestDelegate to unify the load request delegates.  Use didReceiveQueueItemIDs to detect when new media has loaded (external and internal).  Use queueFetchItemsForIDs to ensure that the media items data does exist before calling the loadMediaCallback.  Remove unused remoteMediaClientListener events.

Lindsay-Needs-Sleep committed:
(ios) WIP Use lastMedia to keep track of when the next item in a queue begins playing and to send a manual MEDIA_UPDATE event showing the previous media as finished

Lindsay-Needs-Sleep committed:
(ios) WIP Switch to using isQueueJumping to determine the reason for the queue advancing to another item

Lindsay-Needs-Sleep committed:
(ios) WIP Allow storing multiple callbacks for when the session ends.  Also don't call the endSessionCallbacks until the session has actually ended (but before the SESSION_UPDATE event).

Lindsay-Needs-Sleep committed:
(ios) WIP Forgot to commit these changes with their relevant commit.  (Sorry, did a huge amout of fixes and then tried to make commits for each section of change, to have a slightly more informative git history.  But it is difficult...)

Lindsay-Needs-Sleep committed:
(ios) WIP fix crash when textTrackStyle.backgroundColor does not exist

Lindsay-Needs-Sleep committed:
(ios) WIP Fix contentID output.  For queues, only contentURL is available, but for loadMedia media, only contentID is available

Lindsay-Needs-Sleep committed:
(ios) WIP bug fix, selectRoute should only return the "Leave/Stop current session before selecting" error message when a current connected session exists

Lindsay-Needs-Sleep committed:
(ios) WIP sometimes initialize -> tryRejoin can happen before didEndCastSession can update/erase the currentSession.  This just provides some additional protection against that.

Lindsay-Needs-Sleep committed:
(ios) WIP For some reason didResumeCastSession is called randomly even once we already have the session. (Maybe it temporarily/very quickly loses and regains connection).  Whatever the reason, if we already  have the session, we shouldn't trigger the SESSION_LISTENER event. (Also no point in "setSession" either, since it is the same session.

Lindsay-Needs-Sleep committed:
(ios) WIP detect join session fail and return an error

Lindsay-Needs-Sleep committed:
(test) WIP fix test "ueue should start the next item automatically when previous one finishes" so that it does not re-use var i (causing issue if we need to hit the media listener more than once).  And make it so that the current play time of the item must be close to it's start time once it is reported as playing.

Lindsay-Needs-Sleep committed:
(test) WIP fix copypasta

Lindsay-Needs-Sleep committed:
(test) WIP allow requestSession's stop casting option to trigger the update listener and the success/cancel event in any order.  (Potentially it can happen in either order for chrome desktop since the success/cancel event happens when the casting dialog hides.  Theoretically it is possible to have the dialog hide before the session ends.)

Lindsay-Needs-Sleep committed:
(test) WIP auto tests should have a before which ensures that initialization has occurred so that we can use ".only"

Lindsay-Needs-Sleep committed:
(test) WIP wrap media functions in a describe to improve ".only" use for testing just media functions

Lindsay-Needs-Sleep committed:
[no change] formatting. Change indent for tests required from last 2 commits.

Lindsay-Needs-Sleep committed:
[no change] satisfy eslint and miniscule performance boost

Lindsay-Needs-Sleep committed:
(ios) Have the plugin start onload.   This is supposed to be required to make session resume after app restart work correctly.

Lindsay-Needs-Sleep committed:
(ios) Fix issue https://github.com/miloproductionsinc/cordova-plugin-chromecast/issues/5  Added tests for the issue.

Lindsay-Needs-Sleep committed:
(ios) [no change] remove unused function

Lindsay-Needs-Sleep committed:
(test) format the test homepage to look a bit nicer

Lindsay-Needs-Sleep committed:
(test) Do not require the receiver updates for the before initialize (because chrome broswer will not send receiver updates on the second call), also it is nice for the before to be simple since it is not meant to be testing initialize.

Lindsay-Needs-Sleep committed:
(ios) Fix communication problem with Android.  Android still requires the contentID to be set on media items (ios says it's deprecated), but without it, android counts ios loaded media messages to be malformed.  Maybe Android needs to update the google cast library.  Also removed some code duplication.   Had 2 methods that were both building almost identical MediaInformation items.

Lindsay-Needs-Sleep committed:
(test) allow media after resume to be buffering or playing

Lindsay-Needs-Sleep committed:
(android) Improve/simplify queue item handling

Lindsay-Needs-Sleep commented:
Woo.  ~40 more commits to ios and I think it is ready to go again!

anthonylavado commented:
> * Upgraded to CAF v3, this got rid of a huge amount of deprecation notices.

https://github.com/jellyfin/jellyfin-chromecast 😟

Lindsay-Needs-Sleep commented:
Have you tried it using this update with your receiver?  If jellyfin is at least on a v2 receiver it should work (apparently).

From: https://developers.google.com/cast/docs/caf_receiver

> Note: The Cast Application Framework Receiver SDK is compatible with both v2 and v3 senders.

anthonylavado commented:
>Have you tried it using this update with your receiver?  If jellyfin is at least on a v2 receiver it should work (apparently). ...

@Lindsay-Needs-Sleep ah we do have a V2 receiver, so we’ll test it 🙂

Lindsay-Needs-Sleep committed:
(android) Add a try catch on initialization so that devices incapable of using chromecast do not crash when this plugin is included

Lindsay-Needs-Sleep committed:
(test) Change audioUrl because old one is gone.  Fix readme typo.

Lindsay-Needs-Sleep committed:
Improve queueReloadCallback handling.  Unset it when an error occurs.  Only set it for external loads that have more than 0 items.

Lindsay-Needs-Sleep committed:
(android) remove annoying e.printStackTrace's for invalid (but optional) JSON parsing when creating a MediaInfo object

Lindsay-Needs-Sleep committed:
Added 2 tests:
1) Ensure that we can get media updates after automatically rejoining a session after page reload and app restart (test_manual_primary_1.js) (second part of https://github.com/miloproductionsinc/cordova-plugin-chromecast/issues/5)
2) Ensure that you can't issue media control commands to an outdated media object after a new media object has been loaded (test_auto.js)

To fix #1: (chrome.cast.js) _currentMedia object was removed to ensure that we only send mediaUpdate events to the media stored in the session object.  All media object management is handle through the session object as well now.

To fix #2: (chrome.cast.js) we make sure to invalidate the media.mediaSessionId whenever new media is detected.  For chrome desktop, the new media will always have a different mediaSessionId, but for us, (on android at least), newly loaded media will have the same mediaSessionId, hence the invalidation of the previous media's mediaSessionId.  Additionally, in the case of new media, the old media object stored in session is removed and a new one added (instead of just updating like we were doing previously).

Lindsay-Needs-Sleep committed:
(ios) build returned objects line-by-line instead of all at once to ease debugging (better stack traces)

Lindsay-Needs-Sleep committed:
Handle invalid application receiver ids better

Lindsay-Needs-Sleep committed:
(android) queues should only load 2-3 items into media.items.  If current item index is i, you load items [i-1, i, i+1] *exclude i-1 if it is < 0, and exclude i+1 if it is >= queue.length.  (aka. don't wrap the items.)  This is to match chrome desktop behavior.

Fixes a bug where larger queues (maybe 20+) result in a loop and queueLoad never returns.  This is also safer because it should not result in a crash due to eating all the memory for extremely large queues.

Lindsay-Needs-Sleep committed:
(test) Improve duration checking

dkanada commented:
@Lindsay-Needs-Sleep as you can see no one has had time to review this pull request yet. One of the main reasons is probably that this repository isn't directly related to Jellyfin, so no one has been very interested in maintaining it on our team. Would you be interested in taking ownership of the plugin somewhere? We can work out the details if that sounds okay to you.

Lindsay-Needs-Sleep committed:
(ios) fix orderId on queue itmes

Lindsay-Needs-Sleep committed:
Improve error output from waitForAllCalls

Lindsay-Needs-Sleep committed:
[no change] just a little cleanup, return nothing instead of return false for successfull preChecks (chrome.cast.js)

Lindsay-Needs-Sleep committed:
(ios) Fix https://github.com/miloproductionsinc/cordova-plugin-chromecast/issues/6  This should have been fixed with https://github.com/miloproductionsinc/cordova-plugin-chromecast/issues/5 but the test that covered this case was effectively skipped.  So the test now works properly and you can control the media after an app restart.

Lindsay-Needs-Sleep committed:
Add a little protection from js errors if a media update event is sent before the client has received the session.  (Can happen on occaision when resuming a session on page reload or app restart).

Lindsay-Needs-Sleep committed:
(ios) move retry utility function to CastUtilities

Lindsay-Needs-Sleep committed:
(test) should have a timeout for this test

Lindsay-Needs-Sleep committed:
(ios) Fixes https://github.com/miloproductionsinc/cordova-plugin-chromecast/issues/8
ios does not have mediaStatus on didResumeCastSession if initialize is called immediately after app restart.  So retry for 2 seconds to see if we can get any media.  If not, we will assume the session does not have any media loaded.

Lindsay-Needs-Sleep committed:
(ios) remove NSLog's

Lindsay-Needs-Sleep commented:
@dkanada  Sure.  I mean, I won't be able to dedicate much more time to this project (but I *think* it is working pretty good now).  I don't mind if we "officially" transfer it.  I should be able to answer questions/reviewing PR's.  I'm not sure what else ownership entails?

danielro commented:
Figure this out because right now when one searches for "cordova chromecast plugin" it's not easy to get to Lindsay's plugin and it should.

Lindsay-Needs-Sleep committed:
(ios) (wip) remame objc files to have a namespace

Lindsay-Needs-Sleep committed:
(ios) Fix imports and usage after file + class rename

Lindsay-Needs-Sleep committed:
(ios) improve error handling for loadMediaCallback

Lindsay-Needs-Sleep committed:
Fix https://github.com/miloproductionsinc/cordova-plugin-chromecast/issues/9  Try to keep the media up to date by forcing it to be an empty array whenever media doesn't exist (instead of setting mediaSessionId to 'invalidated').

dkanada commented:
Hmmm since your fork is hard to find it might be easier to just handle the issue here. If you consider your changes almost finished we could make sure the existing master is tagged with a release, then pull in your changes. If you plan on making more changes in the future we could also give you write access to this repository so you can push to master.

Lindsay-Needs-Sleep commented:
Haha, yep.  I thought it was finished basically every commit since I submitted the pull request, but a couple of users kept finding bugs.  Thankfully, the last one was closed yesterday, and no new ones have been reported, so I really, really, think it is done now.

Yep, that sounds like a fine plan!  

I don't have any plans for the immediate future.  Probably #51 at some point.  And maybe adding support for additional functions (eg. chrome.cast.media.QueueInsertItemsRequest because there seems to limit on message size, so you can't create really large queues [~70+] in one go.)  Also, maybe improving the documentation later (divide it up between user documentation and developer documentation).

th3hamm0r commented:
@Lindsay-Needs-Sleep fyi: I've tested your latest master and the crash on iOS (https://github.com/jellyfin/cordova-plugin-chromecast/issues/53) is still happening

anthonylavado commented:
@Lindsay-Needs-Sleep @dkanada Since the current master that we're using is already tagged as a release, I'm willing to:
* Merge these changes
* Add @Lindsay-Needs-Sleep as a maintainer for this repo

Let me know what you think.

Lindsay-Needs-Sleep commented:
Yep, I am fine with that.
I will write a test for [PR 12](https://github.com/miloproductionsinc/cordova-plugin-chromecast/pull/12) in a few days (hopefully) and merge that in as well.  

We can wait for that, or increment the version (1.0.1), or whatever the preferred method is. :)

Lindsay-Needs-Sleep commented:
@anthonylavado Thanks for the merge and invitation! :D

I was wondering if it would be possible to un-squash the commits in the merge?

I was quite meticulous with writing detailed commit messages so that git blame would be useful in determining exactly why and what importance each piece of code does/holds.  I would really hate to lose all of that!  (I know not everyone uses git blame, but I find it super useful!)

anthonylavado commented:
@Lindsay-Needs-Sleep Sorry for the delay - I'm not very good at Git, and I forgot about my GitHub default for squash.

@dkanada Here's a reminder 💯

dkanada commented:
@Lindsay-Needs-Sleep I reset our master branch to just before your pull request was merged. If you open a new PR with the same changes we can merge it again.
